### ISSUE TITLE
ExitHooks run on reload or during the ++ command

### ISSUE THREAD

**Issue by dragos** [sbt/sbt#7268]
## steps

```
Global / onLoad := { state =>
    println("!! Installed onLoad hook.")
    state.addExitHook(println("!! ExitHook called"))
}
```

## problem

Sbt has two different hooks for "end of build": [onUnload](https://github.com/sbt/sbt/blob/6dfebc689b074b0da81c51543914048ffda06e71/main/src/main/scala/sbt/Keys.scala#L90-L91) and [exit hooks](https://github.com/sbt/sbt/blob/d9acfed2205145d972d05a8f7c74ed94b68b7248/main-command/src/main/scala/sbt/State.scala#L44) as part of the Sbt state. It looks like the exit hooks are not different, and run at the same time as unload hooks, making them unsuitable for doing something when Sbt finishes work.

```
❯ sbt
[info] welcome to sbt 1.8.2 (BellSoft Java 1.8.0_342)
[info] loading global plugins from /Users/idragos/.sbt/1.0/plugins
[info] loading settings for project test-sbt-build from plugins.sbt ...
[info] loading project definition from /Users/idragos/sandbox/test-sbt/project
[info] loading settings for project test-sbt from build.sbt ...
[info] set current project to test-sbt (in build file:/Users/idragos/sandbox/test-sbt/)
!! Installed onLoad hook.
[info] sbt server started at local:///Users/idragos/.sbt/1.0/server/616d6fb9d4a3b1397596/sock
[info] started sbt server
sbt:test-sbt> reload
[info] welcome to sbt 1.8.2 (BellSoft Java 1.8.0_342)
!! ExitHook called
[info] loading global plugins from /Users/idragos/.sbt/1.0/plugins
[info] loading settings for project test-sbt-build from plugins.sbt ...
!! ExitHook called
[info] loading project definition from /Users/idragos/sandbox/test-sbt/project
[info] loading settings for project test-sbt from build.sbt ...
!! ExitHook called
[info] set current project to test-sbt (in build file:/Users/idragos/sandbox/test-sbt/)
!! Installed onLoad hook.
sbt:test-sbt> 
```

## expectation
Exit hooks should run only when Sbt really exits.

Any reason for this to stay like this, or a PR fixing it has a good chance to be approved?

## notes

The behavior may seem a minor corner case, but cross compilation and the Switch command (`+`, `++`, `^`, `^^`) all rely on reloading the build and reapplying settings, meaning that exit hooks get invoked even without a direct `reload` call.

**Comment by eed3si9n** [comment]
I think this works as intended. In 8c7efd02b4cd588cbf2c589f539785cc07e73ba8 ScalaDoc comments were added but, unfortunately somewhat inconsistent with itself:

```scala
/**
@param exitHooks code to run before sbt exits, usually to ensure resources are cleaned up.
....
*/
final case class State(
....

/** Convenience methods for State transformations and operations. */
trait StateOps {
	/** Registers a new exit hook, which will run when sbt exits or restarts.*/
	def addExitHook(f: => Unit): State
```

So in this naming convention, `reload` and any command that implicitly calls `reload` conceptually _exits_ an sbt session. This makes sense since settings are supposed to be immutable, and stateful cross building via `++` changes settings.

**Comment by dragos** [comment]
It does seems so, though it is still very confusing. 😢  I think most people associate "sbt exit hooks" with ending a user-facing session, while onUnload hooks are what you described. There is also an sbt `exit` command which does the usual thing, and the scaladoc on `ExitHook` says "/** Defines a function to call as sbt exits.*/".

I am still wondering what is the behavioral difference between the onUnload and exit hooks, since both run at the same time? I understand that one is part of the state and the other part of project settings, which makes the perception that exit hooks are the "outer-most" hooks even stronger.

To summarize this, sbt has several variations of "unload", in increasing order of "depth":

Action|hook|user-facing?
-|-|-
reload|onUnload/ExitHook|yes, can be intercepted in the build definition
reboot|none|not overridable
exit|ShutdownHooks|private to sbt

According to scaladoc in the same commit, ExitHooks are the [alternative to ShutdownHooks](https://github.com/sbt/sbt/commit/8c7efd02b4cd588cbf2c589f539785cc07e73ba8#diff-340e0aee0748b7338cd2954a2fca8e1553cd238b4f73089812701ecd3aafdbbeR53-R59) in case of reboot, which I think makes sense. However, currently, there is no way to react to the middle or last row in a clean way, and exit hooks don't actually cover that (there's no way to know from an exit hook if you are in row 1 or 2).

Shutdown hooks at the JVM level are not a good option (I tried) because they run on a separate thread and lead to race conditions with the sbt shutdown hook thread, leading to crashes when trying to log something (`ChannelClosedException`). This is quite hard to work around, since the sbt terminal seems to replace even `Console.out/err`, so not entirely obvious how to use it.

Since it's probably too late to change the semantics of ExitHooks, I think it'd be good to add a hook for clean shutdown/ reboot that happens before JVM shutdown. Sbt itself could use it, since [this call](https://github.com/sbt/sbt/blob/341e09a07eb374f30b656bffb79ef4adf091f928/main-command/src/main/scala/sbt/internal/server/Server.scala#L161) could eventually fail if the Sbt shutdown hooks are installed in a different order and the global logger is closed when it gets there.

I see two alternatives:
- extend trait `ExitHooks` with a boolean parameter, something along the lines `def runBeforeExiting(terminating: Boolean): Unit`, and set the boolean to `true` when truly terminating. This can be done in a backwards-compatible way, where existing exit hooks keep working as today
- add another type of hook to `State`, say `RebotHooks` and/or `TerminateHooks` to cover the other two rows in the table.

I can't think of another alternative, WDYT?

**Comment by eed3si9n** [comment]
Could you elaborate the underlying motivation for the JVM shutdown?

My guess is that the choice that we don't provide hooks into JVM shutdown was also by-design, because we want the users to have have as fresh sbt session as possible upon `reload`. For example, if there's a development server, and the server's port number is controlled via a global setting, we would like the dev server to be shutdown upon reload, and restarted with the potentially new port number.

In other words, if something goes wrong in the plugin, the user should be able to get back to a fresh state by running `reload`. If we gave further shutdown hooks, then the user might also need to perform JVM shutdown, which might have other consequences like BSP clients getting disturbed.

**Comment by dragos** [comment]
Sure, I am working on "build scans" for sbt, which are a complete record of what happens during a build. On the CI, it is common to pass several commands on the command line, for example `++test publish`, and we want to publish a single build scan that covers one CI run. I need a way to know that Sbt has finished execution and is about to shut down. Sbt itself is using a shutdown hook for printing the "shutting down build server" message, so that's a second usecase where the lifetime of a component is not tied to the "loaded session" lifetime.

My understanding from git archeology and existing scaladoc comments is that the existing `ExitHooks` were designed for this use-case, unfortunately it was designed before cross and switch were added to Sbt (as far as I can tell) which make the reload behavior a lot more common (and surprising) than before.

I understand your concern, but in my experience I find it very common to re-start sbt when something is off, and I found reload to not always work. One case is during development of a plugin, where reload doesn't always pick up the changes.

**Comment by eed3si9n** [comment]
It's true that sbt server might need to shutdown at the end of JVM session. But the very existence of sbt server basically is to avoid *both* sbt reload and JVM shutdown to invoke commands, either on a laptop or on CI. For example our main CI YAML job is written like this:

```bash
        ./sbt -v --client mimaReportBinaryIssues
        ./sbt -v --client javafmtCheck
        ./sbt -v --client "Test/javafmtCheck"
        ./sbt -v --client scalafmtCheckAll
        ./sbt -v --client scalafmtSbtCheck
        ./sbt -v --client serverTestProj/scalafmtCheckAll
        ./sbt -v --client headerCheck
        ./sbt -v --client "Test/headerCheck"
        ./sbt -v --client "Test/compile"
        ./sbt -v --client publishLocal
        ./sbt -v --client test
        ./sbt -v --client "serverTestProj/test"
        ./sbt -v --client doc
        ./sbt -v --client "all $UTIL_TESTS"
```

On a GitHub Action, the sbt server will eventually be shutdown, but on a Jenkins machine, I'd expect the agent machine to keep the sbt server alive. Given that sbtn supports all commands already, it's within realm of possibility that sbtn (`sbt --client`) mode becomes the default in sbt 2.x or even during sbt 1.x.

For the purpose of build scan, I wonder if we should extend https://github.com/sbt/sbt/blob/1.9.x/tasks/src/main/scala/sbt/ExecuteProgress.scala to encode an abstraction of JVM shutdown, like a `endJob` event, which will be called either by JVM shutdown or user typing in `endJob`.
